fairlearn/postprocessing/_tradeoff_curve_utilities.py

```diff
@@ -169,50 +175,44 @@ def _interpolate_curve(
     result : :class:`pandas:pandas.DataFrame`
         DataFrame with the points of the interpolated curve.
     """
-    data_transpose = data.transpose()
+    x_values = data[x_col].values
+    y_values = data[y_col].values
+    content_values = data[content_col].values
 
     content_col_0 = content_col + "0"
     content_col_1 = content_col + "1"
 
-    i = 0
-    dict_list = []
-    x0 = data_transpose[0][x_col]
-    while data_transpose[i + 1][x_col] == x0:
-        i += 1
-
-    # calculate the curve points for each x tick in x_grid
-    for x in x_grid:
-        # skip over data points that we've already passed
-        while x > data_transpose[i + 1][x_col]:
-            i += 1
-
-        # Calculate the y value at x based on the slope between data points i and i + 1
-        x_distance_from_next_data_point = data_transpose[i + 1][x_col] - x
-        x_distance_between_data_points = data_transpose[i + 1][x_col] - data_transpose[i][x_col]
-        p0 = x_distance_from_next_data_point / x_distance_between_data_points
-        p1 = 1 - p0
-        y = p0 * data_transpose[i][y_col] + p1 * data_transpose[i + 1][y_col]
-        dict_list.append(
-            {
-                x_col: x,
-                y_col: y,
-                P0_KEY: p0,
-                content_col_0: data_transpose[i][content_col],
-                P1_KEY: p1,
-                content_col_1: data_transpose[i + 1][content_col],
-            }
-        )
+    indices = np.searchsorted(x_values, x_grid, side="right") - 1
+    indices = np.clip(indices, 0, len(x_values) - 2)
```

Comment:
To handle the last element in `x_grid` (if it's equal to the biggest element in `x_values`)